hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/blockmanagement/DatanodeManager.java

```diff
@@ -1856,15 +1875,16 @@ public DatanodeCommand[] handleHeartbeat(DatanodeRegistration nodeReg,
       nodeinfo.setBalancerBandwidth(0);
     }
 
-    if (slowPeerTracker != null) {
-      final Map<String, Double> slowPeersMap = slowPeers.getSlowPeers();
-      if (!slowPeersMap.isEmpty()) {
-        if (LOG.isDebugEnabled()) {
-          LOG.debug("DataNode " + nodeReg + " reported slow peers: " +
-              slowPeersMap);
-        }
-        for (String slowNodeId : slowPeersMap.keySet()) {
-          slowPeerTracker.addReport(slowNodeId, nodeReg.getIpcAddr(false));
+    synchronized (slowPeerTrackerLock) {
```

Comment:
Thanks @virajjasani for updating. Adding lock `slowPeerTrackerLock` looks good. But I'm not sure if it affects heartbeat here.

I have one small suggestion:
1. Make dataNodePeerStatsEnabled volatile.
2. When dataNodePeerStatsEnabledVal = false, do not set slowPeerTracker to null.
3. Change the condition from `if (slowPeerTracker != null)` to `if (slowPeerTracker != null) && dataNodePeerStatsEnabled` where `slowPeerTracker` is used.

Disadvantages:
if `dataNodePeerStatsEnabledVal = false`, `slowPeerTracker` will be retained. But I guess the operation to `disable slowPeerTracker` is not common.

Advantages:
Does not cause lock contention and does not cause NPE.

What do you think of this?

Reply:
Will get back with adding new class `SlowPeerDisabledTracker` that extends `SlowPeerTracker` such that all extended methods in SlowPeerDisabledTracker are no-op, hence we never need to assign null to `slowPeerTracker` of DatanodeManager in the first place and hence never need to deal with NPEs. This will save us from using volatile or synchronization as well because both of them would add to the performance of heartbeat handling.

Toggling b/ enabled and disabled mode of "slow peer tracking" would be much simplified this way. Let me get back with the changes tomorrow.

Thanks @tomscut for the nice discussion!